### Title : First run ML Flow - Hello World

- In this project, i will run the mlflow- hello world program to get the basic idea on things

In [33]:
import mlflow, mlflow.sklearn
from sklearn.datasets import load_iris
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [34]:
#load iris data.
x, y = load_iris(return_X_y=True)

#split training data into 80-20 split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

print(f'Train data size: {X_train.shape}, Test data size: {X_test.shape}')


Train data size: (120, 4), Test data size: (30, 4)


#Mental Model
┌─────────────────────┐
│  150 iris flowers   │
└──────────┬──────────┘
           │ train_test_split(test_size=0.2, random_state=42)
           ▼
   ┌───────┴────────┐
   │                │
   ▼                ▼
┌──────┐         ┌──────┐
│ 120  │         │  30  │
│TRAIN │         │ TEST │  ← model never sees these during training
└──┬───┘         └──┬───┘
   │                │
   │ model.fit()    │ model.predict()  → compare to true y_test
   ▼                ▼
 LEARN          GRADE

### ❓ Why don't we evaluate on the training data?

Evaluating on training data does not make sense because the model has already seen this data during training.  
It may simply memorize patterns instead of learning to generalize.

The actual evaluation should be done on **unseen data** (test dataset) to measure how well the model performs in real-world scenarios.

---

### ❓ What does `random_state` do?

`random_state` is used to make results reproducible.

By setting a fixed `random_state` (seed), the same train-test split will be generated every time you run the code.  
This ensures consistency in experiments and debugging.

---

### ❓ What's the shape of `X_train` if Iris has 150 rows and `test_size = 0.2`?

- Total samples = 150  
- Test size = 20% → 30 samples  
- Train size = 150 - 30 = 120 samples  
X_train.shape = (120, 4)

In [ ]:
mlflow.set_experiment("iris-hello-world")
with mlflow.start_run():
    c = 95
    max_iter = 200
    acc = 0
    model = LogisticRegression(C=c,max_iter=max_iter)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    accuracy_score = accuracy_score(y_test, y_pred)
    mlflow.log_param("C", c)
    mlflow.log_metric("accuracy", accuracy_score)
    mlflow.sklearn.log_model(model, "model")
    print(f'accuracy_score : {accuracy_score}')

2026/04/26 16:22:13 WARNING mlflow.models.model: `artifact_path` is deprecated. Please use `name` instead.
2026/04/26 16:22:13 WARNING mlflow.sklearn: Saving scikit-learn models in the pickle or cloudpickle format requires exercising caution because these formats rely on Python's object serialization mechanism, which can execute arbitrary code during deserialization. The recommended safe alternative is the 'skops' format. For more information, see: https://scikit-learn.org/stable/model_persistence.html
2026/04/26 16:22:16 WARNING mlflow.utils.environment: Failed to resolve installed pip version. ``pip`` will be added to conda.yaml environment spec without a version specifier.


accuracy_score : 0.9333333333333333
